src/editor/classes/UiDriver.js

```diff
@@ -1,31 +1,153 @@
+        if(prevReturn !== undefined) {
+            this.setReturnData(prevReturn);
+        }
+        return prevReturn;
+    }
+
+    // BEGIN HOOKS
+    // These functions hook into the CodeMirror interface and allow us to
+    // perform asynchronous data transfer between C++ and Javascript.
+
+    this.onOptionChange = function(editor) {
+        var indentMode = [editor.options.indentWithTabs, editor.options.indentUnit];
+        this.proxy.setValue("indentMode", indentMode);
+    }
+
+    // Hook for when the editor's language is changed.
+    this.onLanguageChange = function(editor) {
+        var langId = Languages.currentLanguage(editor);
+        var langData =  {id: langId, lang: Languages.languages[langId]}; 
+        this.proxy.setValue("language", langData);
+        this.proxy.sendEditorEvent("J_EVT_CURRENT_LANGUAGE_CHANGED", langData);
+    }
+
+    // Hook for when we load a file/change content
+    this.onSetValue = function(editor) {
+        var detectedIndent = this.detectIndentationMode(editor);
+        if (detectedIndent !== undefined) {
+            this.proxy.setValue("detectedIndent", detectedIndent);
+        }
+        editor.clearHistory();
+        this.proxy.sendEditorEvent("J_EVT_DOCUMENT_LOADED", 0);
+    }
+
+    // Hook for when cursor activity is detected(cursor moved/selection changed)
+    this.onCursorActivity = function(editor) {
+        var out = [];
+        var sels = editor.listSelections();
+        for (var i = 0; i < sels.length; i++) {
+            out[i] = {  anchor: {
+                            line: sels[i].anchor.line,
+                            ch: sels[i].anchor.ch
+                        },
+                        head: {
+                            line: sels[i].head.line,
+                            ch: sels[i].head.ch
+                        }
+                    };
+        }
+        this.proxy.setValue("selections", out);
+        this.proxy.setValue("selectionsText", editor.getSelections("\n"));
+        var cur = editor.getCursor();
+        this.proxy.setValue("cursor", [cur.line, cur.ch]);
+    }
+
+    // Hook for when any content is changed
+    this.onChange = function(editor) {
+        this.proxy.setValue("charCount", editor.getValue("\n").length);
+        this.proxy.setValue("lineCount", editor.lineCount());
+        this.proxy.setValue("clean", isCleanOrForced(changeGeneration));
+    }
+
+    // Hook for when the user scrolls the page.
+    this.onScroll = function(editor) {
+        var scroll = editor.getScrollInfo();
+        this.proxy.setValue("scrollPosition", [scroll.left, scroll.top]);
+    }
+
+    // Hook for when the editor is in a functional loaded state
+    this.onLoad = function(editor) {
+        this.onCursorActivity(editor);
+        this.onChange(editor);
+        this.onScroll(editor);
+        this.onOptionChange(editor);
+        this.proxy.sendEditorEvent("J_EVT_READY", 0);
+        editor.refresh();
+    }
+    // END HOOKS
+}
+
```

Comment:
`detectIndentationMode`, `onLanguageChange`, `onSetValue`, `onCursorActivity`, `onChange` and `onScroll` shouldn't be here.

`onOptionChange` probably neither.

This file should only contain abstract communication functions and shouldn't be aware of any details about how our Editor works.

Let's discuss a better place to put these functions.